File: web/src/api/danmaku.ts
import { api } from './client'

// DanmakuEpisode / DanmakuAnime mirror the dandanplay search response. When
// multiple anime match, the backend returns them as candidates and the player
// asks the user which to load (disambiguation).
export interface DanmakuEpisode {
  episodeId: number
  episodeTitle: string
}

export interface DanmakuAnime {
  animeId: number
  animeTitle: string
  episodes: DanmakuEpisode[]
}

// DanmakuFetchResult mirrors the backend /api/danmaku/:id response: renderer
// knobs (resolved from runtime settings) plus the raw upstream payload. The
// dandanplay protocol delivers Bilibili-format XML or its own JSON shape
// ({comments:[{p,m,t}]}); source_type is sniffed from the body by the backend
// ("auto" when there is nothing to sniff). When several anime matched,
// candidates is non-empty and raw is empty.
export interface DanmakuFetchResult {
  enabled: boolean
  source_type: 'auto' | 'xml' | 'json'
  opacity: string
  font_size: string
  area: string
  raw?: string
  candidates?: DanmakuAnime[]
}

export type DanmakuFetchOptions = {
  /** Overrides the media-derived search keyword (manual search). */
  kw?: string
  /** Forces a specific danmaku library chosen by the user. */
  episodeId?: number | string
}

// danmakuAPI fetches danmaku comments for a media item. The backend resolves
// the configured dandanplay source by the video's name (search for episode,
// then fetch its comment library) and returns the raw Bilibili-format XML;
// parsing into comment objects happens client-side in utils/parseDanmaku.
export const danmakuAPI = {
  fetch: (mediaId: string, options: DanmakuFetchOptions = {}) =>
    api
      .get<DanmakuFetchResult>(`/danmaku/${encodeURIComponent(mediaId)}`, {
        params: { kw: options.kw || undefined, episodeId: options.episodeId || undefined },
        timeout: 20_000,
      })
      .then((r) => r.data),

  // config returns the renderer knobs so the player can initialize its
  // danmaku control panel without admin privileges.
  config: () =>
    api
      .get<{
        enabled: boolean
        source?: string
        opacity: string
        font_size: string
        area: string
      }>('/danmaku/config')
      .then((r) => r.data),
}

File: internal/handler/danmaku.go
// Package handler — danmaku endpoints.
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"github.com/truewhile/MeBox/internal/service"
)

// getDanmakuHandler returns danmaku for a media. ?kw= optionally overrides the
// search keyword (used by the player's manual search box); ?episodeId= forces a
// specific danmaku library chosen by the user after a disambiguation.
func getDanmakuHandler(svc *service.Container) gin.HandlerFunc {
	return func(c *gin.Context) {
		res, err := svc.Danmaku.Fetch(c.Request.Context(), c.Param("id"), c.Query("kw"), c.Query("episodeId"))
		if err != nil {
			c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, res)
	}
}

// getDanmakuConfigHandler exposes the danmaku renderer knobs (opacity, font
// size, area, enabled) so the player can initialize its control panel without
// admin privileges.
func getDanmakuConfigHandler(svc *service.Container) gin.HandlerFunc {
	return func(c *gin.Context) {
		c.JSON(http.StatusOK, svc.Danmaku.Config(c.Request.Context()))
	}
}
